gpflow/base.py

```diff
@@ -320,13 +344,15 @@ def _cast_to_dtype(value: VariableData, dtype: Optional[DType] = None) -> tf.Ten
         return tf.convert_to_tensor(value, dtype=dtype)
 
 
-def _to_constrained(value: VariableData, transform: Transform) -> tf.Tensor:
+def _to_constrained(value: tf.Tensor, transform: Optional[Transform]) -> tf.Tensor:
```

Comment:
TensorLike not Tensor

Reply:
this is a tricky one. i thought very carefully about these types. if we make this anything other than `tf.Tensor`, there's not guarantee `value` and `read_value` on `Parameter` will return the types required for a `tf.Variable`

Reply:
I could allow `TensorLike` by using overloads, but that's overkill for a private, essentially one-line, function

Reply:
It's not correct to say that this function requires `value` to be of type `tf.Tensor`. I'd rather not have a type hint than one that's incorrect - so given that it's private, maybe just remove it ?

Reply:
i don't think it's true to say it's not correct. it's perfectly OK to ask for a stricter type than the function implementation supports. then again, if you don't want us to do that in this case, that's a separate valid question

Reply:
i think typehints specify the api further and agree we shouldn't tighten if we intend people to be using tensor like. i agree with ti's second suggestion, since private.

Reply:
by remove it, do you mean remove the function or remove the types?

Reply:
if you mean remove the types, my concern is we're losing out on this part of the original ticket #1368 
> As a GPflow maintainer, I want full type hint coverage so that I can be confident my changes are sound.

Reply:
that point is of course up for debate, but I want to draw attention to it

Reply:
how about I use `tf.convert_to_tensor` at the call site instead?

Reply:
> As a GPflow maintainer, I want full type hint coverage so that I can be confident my changes are sound.

As a GPflow maintainer, I don't want "full type hint coverage" to come at arbitrary cost (w.r.t. type restrictions, code readability, and others).

> how about I use tf.convert_to_tensor at the call site and use TensorType here?

Please don't change the behaviour just to add types.

Reply:
ok. sounds like "full type coverage" isn't quite what we're aiming for. I guess it's not really doable in python anyway. I'll have a think about what we can aim for.

my assumption (ok, more a hope) was that it wouldn't change the functionality since it's returning a Tensor anyway. but I accept there's a risk it might, so I'll rethink that

Reply:
I've gone with typing it to `TensorType` cos I think that's better than no type. I've put `# type: ignore` at call site